gcc/rust/hir/rust-ast-lower-type.cc

```diff
@@ -445,17 +446,28 @@ ASTLowerGenericParam::translate (AST::GenericParam &param)
 void
 ASTLowerGenericParam::visit (AST::LifetimeParam &param)
 {
+  AST::Attribute outer_attr = AST::Attribute::create_empty ();
   auto crate_num = mappings.get_current_crate ();
+  AST::Lifetime lifetime = param.get_lifetime ();
   Analysis::NodeMapping mapping (crate_num, param.get_node_id (),
 				 mappings.get_next_hir_id (crate_num),
 				 mappings.get_next_localdef_id (crate_num));
 
-  HIR::Lifetime lt (mapping, param.get_lifetime ().get_lifetime_type (),
-		    param.get_lifetime ().get_lifetime_name (),
-		    param.get_lifetime ().get_locus ());
+  for (auto &o_attr : param.get_outer_attrs ())
+    {
+      rust_assert (!o_attr.is_inner_attribute ());
+
+      Rust::AST::SimplePath path = o_attr.get_path ();
+      if (path == Values::Attributes::MAY_DANGLE)
+	outer_attr = o_attr;
+    }
+
+  HIR::Lifetime lt (mapping, lifetime.get_lifetime_type (),
+		    lifetime.get_lifetime_name (), lifetime.get_locus ());
 
-  translated = new HIR::LifetimeParam (mapping, lt, param.get_locus (),
-				       std::vector<Lifetime> ());
+  translated
+    = new HIR::LifetimeParam (mapping, lt, param.get_locus (),
+			      std::vector<Lifetime> (), std::move (outer_attr));
```

Comment:
hm, I'm not sure I understand the purpose of this code. we only set an outer attribute to lifetime parameters if they have a `#[may_dangle]` one? can they not have multiple attributes? I don't know exactly how attributes work on lifetimes so I'm genuinely curious haha

Reply:
This was because of a mismatch between AST and HIR. AST has an AttrVec, but HIR has:
```
651 // A lifetime generic parameter (as opposed to a type generic parameter)
652 class LifetimeParam : public GenericParam
653 {
654   Lifetime lifetime;
655
656   // bool has_lifetime_bounds;
657   // LifetimeBounds lifetime_bounds;
658   std::vector<Lifetime> lifetime_bounds; // inlined LifetimeBounds
659
660   // bool has_outer_attribute;
661   // std::unique_ptr<Attribute> outer_attr;
662   AST::Attribute outer_attr;
```

I have absolutely no idea whether a lifetime (or type param) can have multiple attributes. If I had to guess, the answer is yes, but I can't find an attribute other than may_dangle to test it out on. On another note, it seems like rust is doing the pass for may_dangle on the AST instead of in HIR.

Reply:
The best evidence I have that rustc supports multiple attributes is this:
```
liam@gentoo ~/dangle $ rustc  main2.rs
error[E0517]: attribute should be applied to a struct, enum, or union
 --> main2.rs:7:20
  |
7 | unsafe impl<#[repr(C)] #[may_dangle] 'a> Action for Inspector2<'a> {}
  |                    ^                 -- not a struct, enum, or union

error: aborting due to 1 previous error
```
Note, it thinks repr(C) is applied to 'a not #[may_dangle])

Reply:
I'll write up a patch for this soon, it's a reasonable change to make.

Reply:
Fixing the AttrVec stuff was somewhat involved. I've attached a changeset to make it easier to review

Reply:
[0001-squash-and-post-later.patch](https://github.com/user-attachments/files/16312084/0001-squash-and-post-later.patch)